File: generate/lib/RunHugo.js
var async = require('async');
var util = require('util');
var spawn = require('child_process').spawn;
var s3 = require('s3');
var AWS = require('aws-sdk');

var syncClient = s3.createClient({
    maxAsyncS3: 20,
});

tmpDir = "/tmp/sources";
pubDir = tmpDir + "/public";

function isDir(dirName) {
    var isDirRe = new RegExp(/\/$/);
    return dirName.match(isDirRe) !== null;
}

function isStatic(fName) {
    var isStaticRe = new RegExp(/(^static\/|\/static\/)/);
    return fName.match(isStaticRe) !== null;
}

function staticFile(srcBucket, srcKey, dstBucket, context) {
    var awsS3 = new AWS.S3();
    var dst = srcKey.substring(7);
    var keyMatch = srcKey.match(/\/static\//);
    if (keyMatch !== null) {
        console.log("Key " + srcKey + " is in a theme content directory, removing prefix");
        dst = srcKey.substring(keyMatch.index + 8);
    }
    console.log("Dest: " + dst);
    awsS3.headObject({
        Bucket: srcBucket,
        Key: srcKey,
    }, function(err, data) {
        if (err) { // an error occurred
            console.log("Obj not found");
            context.done(err);
        }
        console.log("Obj has etag: " + data.ETag);
        awsS3.copyObject({
            ACL: 'public-read',
            Bucket: dstBucket,
            Key: dst,
            CopySource: srcBucket + '/' + srcKey,
            CopySourceIfNoneMatch: '"f482b82df157bee673b36145f9641005"',
            StorageClass: 'REDUCED_REDUNDANCY',
        }, function(err, data){

            if (err) { // an error occurred
                if (err.code === "PreconditionFailed") {
                    console.log("Static object already exists in S3 and is unmodified.")
                    context.done();
                } else {
                    console.log(util.inspect(err, {depth: 5}));
                    console.log(err.stack);
                    context.done(err);
                }
            } else {
                context.done();
            }
        });
    });
}

function siteGenerate(srcBucket, srcKey, dstBucket, context) {
    async.waterfall([
    function download(next) {
        var params = {
            localDir: tmpDir,

            s3Params: {
                Bucket: srcBucket,
            },
            getS3Params: function(localfile, s3Object, callback) {
                // skip static content
                if (isStatic(s3Object.Key)) {
                    callback(null, null);
                    return;
                }
                if (isDir(s3Object.Key)) {
                    callback(null, null);
                    return;
                }
                callback(null, {});
            },
        };
        var downloader = syncClient.downloadDir(params);
        downloader.on('error', function(err) {
            console.error("unable to sync down:", err.stack);
            next(err);
        });
        downloader.on('end', function() {
            console.log("done downloading");
            next(null);
        });
    },
    function runHugo(next) {
        console.log("Running hugo");
        var child = spawn("./hugo", ["-v", "--source=" + tmpDir, "--destination=" + pubDir], {});
        child.stdout.on('data', function (data) {
            console.log('hugo-stdout: ' + data);
        });
        child.stderr.on('data', function (data) {
            console.log('hugo-stderr: ' + data);
        });
        child.on('error', function(err) {
            console.log("hugo failed with error: " + err);
            next(err);
        });
        child.on('close', function(code) {
            console.log("hugo exited with code: " + code);
            next(null);
        });
    },
    function upload(next) {
        var params = {
            localDir: pubDir,
            deleteRemoved: false,
            s3Params: {
                ACL: 'public-read',
                Bucket: dstBucket,
            },
        };
        var uploader = syncClient.uploadDir(params);
        uploader.on('error', function(err) {
            console.error("unable to sync up:", err.stack);
            next(err);
        });
        uploader.on('end', function() {
            console.log("done uploading");
            next(null);
        });
    },
    ], function(err) {
        if (err) console.error("Failure because of: " + err)
        else console.log("All methods in waterfall succeeded.");

        context.done();
    });
}

function handleFile(srcBucket, srcKey, dstBucket, context) {
    // bail for .git files
    if (srcKey.match(/^\.git\//) !== null) {
        context.done();
        return;
    }
    if (isDir(srcKey)) {
        context.done();
        return;
    }
    if (isStatic(srcKey)) {
        staticFile(srcBucket, srcKey, dstBucket, context);
    } else {
        siteGenerate(srcBucket, srcKey, dstBucket, context);
    }
}

exports.handler = function(event, context) {
    // Read options from the event.
    console.log("Reading options from event:\n", util.inspect(event, {depth: 5}));
    //var message = JSON.parse(event.Records[0]);
    var srcBucket = event.Records[0].s3.bucket.name;
    var srcKey    = event.Records[0].s3.object.key;
    var dstBucket = event.Records[0].s3.bucket.name.replace('source.', '');

    // don't run hugo for git files
    if (srcKey.match(/^\.git\//) !== null) {
        console.log("Key " + srcKey + " is static content, bailing out");
        context.done();
    }
    handleFile(srcBucket, srcKey, dstBucket, context);
};

File: generate/lib/run_hugo.py
import boto3
import subprocess
import botocore.exceptions
import logging

logger = logging.getLogger()
logger.setLevel(logging.INFO)

s3 = boto.client('s3')

tmp_dir = '/tmp/sources'
dest_dir = '/tmp/sources/public'

def static_file(key, source_bucket, dest_bucket):
    dest_key = key.split('static/')[-1]
    dest_metadata = s3.head_object(Bucket=dest_bucket, Key=dest_key)

    try:
        s3.copy_object(
            ACL='public-read',
            Bucket=dest_bucket,
            Key=dest_key,
            CopySource='/'.join(source_bucket, key),
            # don't copy if the destination object is already correct
            CopySourceIfNoneMatch=dest_metadata['ETag'],
        )
    except botocore.exceptions.ClientError as exc:
        if "PreconditionFailed" not in exc.message:
            logger.error("Failure copying static file")
            logger.exception(exc)


def generate_site(source_bucket, dest_bucket):
    # download files to TMP
    hugo = subprocess.Popen(
        ["./hugo", '-v', '--source=' + tmp_dir, '--destination' + dest_dir],
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
    )
    stdout, stderr = hugo.communicate()

    if hugo.returncode != 0:
        logger.error("Hugo exited with code %d" % exit)
        logger.error("---stdout---\r\n%s\r\n---end stdout---" % stdout)
        logger.error("---stderr---\r\n%s\r\n---end stderr---" % stderr)
    else:
        logger.debug("Hugo exited with code %d" % exit)
        logger.debug("---stdout---\r\n%s\r\n---end stdout---" % stdout)
        logger.debug("---stderr---\r\n%s\r\n---end stderr---" % stderr)

    # upload to cloud


def handler(event, context):
    key = event['Records']['s3']['object']['key']
    source_bucket = event['Records']['s3']['bucket']['name']
    dest_bucket = "source." + source_bucket

    if key.startswith('.git/'):
        logger.debug("Git file, skipping.")
        return
    if key.endswith('/'):
        logger.debug("Directory, skipping.")
        return
    if key.startswith('static/') or '/static/' in key:
        # handle as a static file
        static_file(key, source_bucket, dest_bucket)
    else:
        pass
        # handle default case of generating the whole site
